machinery-helper/machinery_helper.go

```diff
@@ -0,0 +1,231 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.   See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program; if not, contact SUSE LLC.
+//
+// To contact SUSE about this file by physical or electronic mail,
+// you may find current contact information at www.suse.com
+
+package main
+
+import (
+  "os/exec"
+  "bytes"
+  "log"
+  "strings"
+  "io/ioutil"
+  "encoding/json"
+  "os"
+  "sort"
+  "fmt"
+  "path/filepath"
+  "unicode/utf8"
+  "flag"
+)
+
+var Ignore_list map[string]bool
+var RemoteMount_list map[string]bool
+
+func getRpms() []string {
+  cmd := exec.Command("rpm", "-qlav")
+  var out bytes.Buffer
+  cmd.Stdout = &out
+  err := cmd.Run()
+  if err != nil {
+    log.Fatal(err)
+  }
+
+  f := func(c rune) bool {
+    return c == '\n'
+  }
+  packages := strings.FieldsFunc(out.String(), f)
+  return packages
+}
+
+func parseRpmLine(line string) (file_type string, file_name string, link_target string) {
+  file_type = line[0:1]
+
+  index := strings.Index(line, "/")
+  if index < 0 {
+    panic(line)
+  }
+  file := line[index:]
+  fields := strings.Split(file, " -> ")
+  if len(fields) == 2 {
+    file_name = fields[0]
+    link_target = fields[1]
+  } else {
+    file_name = file
+  }
+  return
+}
+
+func addImplicitlyManagedDirs(dirs map[string]bool, files map[string]string) {
+  for file, target := range files {
+    for i:= 1; i < len(file); i++ {
+      if file[i] == '/' {
+        topdir := file[:i]
+        if _, ok := dirs[topdir]; !ok {
+          dirs[topdir] = false
+        }
+      }
+    }
+
+    if target != "" {
+      if _, ok := dirs[target]; ok {
+        dirs[file] = false
+      }
+    }
+  }
+  return
+}
+
+func getManagedFiles() (map[string]string, map[string]bool) {
+  files := make(map[string]string)
+  dirs := make(map[string]bool)
+
+  for _, pkg := range getRpms() {
+    if pkg != "(contains no files)" {
+      file_type, file_name, link_target := parseRpmLine(pkg)
+      switch file_type {
+        case "-":
+          files[file_name] = ""
+        case "d":
+          dirs[file_name] = true
+        case "l":
+          files[file_name] = link_target
+      }
+    }
+  }
+
+  addImplicitlyManagedDirs(dirs, files)
+
+  return files, dirs
+}
+
+func assembleJson(unmanaged_files_map interface{}) string {
+  json_map := map[string]interface{}{ "extracted": false, "files": unmanaged_files_map }
+  json, _ := json.MarshalIndent(json_map, " ", "  ")
+  return string(json)
+}
+
+var readDir = func(dir string) ([]os.FileInfo, error) {
+  return ioutil.ReadDir(dir)
+}
+
+func hasManagedDirs(dir string, rpm_dirs map[string]bool) bool {
+  for rpm_dir := range rpm_dirs {
+    if strings.HasPrefix(rpm_dir, dir + "/") {
+       return true
+    }
+  }
+  return false
+}
+
+func findUnmanagedFiles(dir string, rpm_files map[string]string, rpm_dirs map[string]bool, unmanaged_files map[string]string) {
```

Comment:
don't use underscores in Go names; func parameter rpm_files should be rpmFiles<br>don't use underscores in Go names; func parameter rpm_dirs should be rpmDirs<br>don't use underscores in Go names; func parameter unmanaged_files should be unmanagedFiles